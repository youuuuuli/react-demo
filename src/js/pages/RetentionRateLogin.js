import React, { useContext, useEffect, useMemo, useState } from 'react';
import {
  Button,
  Form,
  Icon,
  List,
  Loader,
  Menu,
  Popup,
} from 'semantic-ui-react';
import DF from '../constants/date_format.json';
import initPagination from '../constants/InitPagination';
import useRetentionRateLogin, { useRetentionRateLoginTypes } from '../constants/useRetentionRateLogin';
import ApiContext from '../default/ApiContext';
import styles from '../../css/DailyReportRate.module.css';
import { DateRange, Pagination, moment } from '../utils';
import TableOs from '../components/TableOs';
import RetentionRateLoginCharts from './RetentionRateLoginCharts';
import { getUserRetentionRet } from '../response';

/**
 * @typedef SearchFormType 定義搜尋條件
 * @property {'percent' | 'number'} type 資料類型
 * @property {'day' | 'week' | 'month'} period 時間類型
 * @property {string} start_at 開始日期
 * @property {string} end_at 結束日期
 */

const RetentionRateLogin = () => {
  const { tr } = useContext(ApiContext);
  const PeriodMap = useRetentionRateLogin();
  const ReportTypes = useRetentionRateLoginTypes();

  /** @type {SearchFormType} 初始搜尋條件 */
  const initSearchForm = {
    type: Object.keys(ReportTypes)[0],
    // 預設顯示：週(week)
    period: Object.keys(PeriodMap)[1],
    // 預設顯示：上個月1號至7號
    start_at: moment()
      .subtract(1, 'month')
      .startOf('month')
      .subtract(-1, 'days')
      .tz('Etc/GMT+4')
      .startOf('day')
      .format(),
    end_at: moment()
      .subtract(1, 'month')
      .startOf('month')
      .subtract(-7, 'days')
      .tz('Etc/GMT+4')
      .endOf('day')
      .format(),
  };

  /** @type {RetentionRateLoginType} 初始登入留存率資料 */
  const initRetentionRateLogin = {
    ret: { percent: [], number: [] },
    sub_total: { percent: {}, number: {} },
    total: { percent: {}, number: {} },
  };

  // 開啟/關閉說明文字
  const [activeExplanation, setActiveExplanation] = useState(false);
  // 登入留存率資料
  const [data, setData] = useState(initRetentionRateLogin);
  // 下載中
  const [loading, setLoading] = useState(false);
  // 分頁資料
  const [pagination, setPagination] = useState(initPagination);
  // 時間類型
  const [period, setPeriod] = useState(initSearchForm.period);
  // 搜尋條件
  const [searchForm, setSearchForm] = useState(initSearchForm);
  // 資料類型
  const [type, setType] = useState(initSearchForm.type);

  /**
   * 處理週期
   *
   * @param {Event} ev event
   * @param {{ 'data-period': string }} params params
   */
  const handleOnClickPeriod = (ev, { 'data-period': value }) => {
    setSearchForm((prev) => ({ ...prev, period: value }));
  };

  /**
   * 處理註冊日
   *
   * @param {object} day 起始/結束日期
   */
  const handleCreatedDate = (day) => {
    const startAt = day.startDate.tz('Etc/GMT+4').startOf('day');
    const endAt = day.endDate.tz('Etc/GMT+4').endOf('day');

    // 日期僅可選擇(今日-1天) 至 (今日-181天) (因數據只統計至前一日)
    const lastDay = moment()
      .subtract(1, 'days')
      .tz('Etc/GMT+4')
      .endOf('day');

    setSearchForm((prev) => ({
      ...prev,
      start_at: startAt.isAfter(lastDay) ? lastDay.format() : startAt.format(),
      end_at: endAt.isAfter(lastDay) ? lastDay.format() : endAt.format(),
    }));
  };

  /**
   * 處理送出搜尋
   */
  const handleSubmit = () => {
    if (loading) {
      return;
    }

    setPeriod(searchForm.period);
    getRetentionRateLogin({});
  };

  /**
   * 處理開啟/關閉說明文字
   */
  const handleExplanation = () => {
    setActiveExplanation(!activeExplanation);
  };

  /**
   * 處理是否為百分比切換
   *
   * @param {Event} ev event
   * @param {{ 'data-type': 'percent' | 'number' }} params params
   */
  const handleIsPercentChange = (ev, { 'data-type': value }) => {
    setType(value);
  };

  /**
   * 處理表格呈現資料
   *
   * @param {object} params params
   * @param {RetentionRateLoginType} params.tableData 登入留存率資料
   * @returns {Array<object>} TableData
   */
  const handleTableData = ({ tableData = data }) => {
    const { ret, sub_total: subTotal, total } = tableData;
    const isPercent = type === 'percent';
    const newRet = ret[type];

    const newData = newRet
      .map((r) => {
        const newR = r;

        newR.date = `${r.day} ${moment(r.day).format('dddd')}`;
        newR.isPercent = isPercent;
        newR.isResultCell = false;
        newR.toColorClassName = isPercent;

        return newR;
      })
      .concat([
        {
          ...subTotal[type],
          date: tr(isPercent ? 'M_TEXT_SUB_TOTAL_AVERAGE' : 'M_TEXT_RESULT_SUB_TOTAL_AMOUNT'),
          isPercent,
          isResultCell: true,
        },
        {
          ...total[type],
          date: tr(isPercent ? 'M_TEXT_AMOUNT_AVERAGE' : 'M_TEXT_RESULT_TOTAL_AMOUNT'),
          isPercent,
          isResultCell: true,
        },
      ]);

    return newData;
  };

  /**
   * 取得登入留存率資料
   *
   * @param {object} params params
   * @param {SearchFormType} params.searchForm 搜尋條件
   * @param {PaginationType} params.pagination 頁碼資訊
   */
  const getRetentionRateLogin = async ({
    searchForm: localSearchForm = searchForm,
    pagination: localPagination = initPagination,
  }) => {
    setLoading(true);

    const out = { result: 'ok', ...getUserRetentionRet[localSearchForm.period] };

    if (out && out.result === 'ok') {
      setData({
        ret: !Object.keys(out.ret).length ? initRetentionRateLogin.ret : out.ret,
        sub_total: out.sub_total,
        total: out.total,
      });
      setPagination(out.pagination);
    }

    setLoading(false);
  };

  /**
   * 登入留存率說明圖示(上下箭頭)
   */
  const useActiveIcon = useMemo(() => (
    activeExplanation ? 'up' : 'down'
  ), [activeExplanation]);

  /**
   * 表格資料
   */
  const useTableData = useMemo(() => {
    if (data.ret[type].length > 0) {
      return handleTableData({});
    }

    return [];
  }, [data.ret, type]);

  useEffect(() => {
    getRetentionRateLogin({});
  }, []);

  return (
    <>
      <div className="page-title">
        {tr('M_TEXT_RETENTION_RATE_LOGIN')}
      </div>

      <Form className="search-wrap">
        <Form.Group>
          {/* 留存率週期 */}
          <Form.Field>
            <label>
              {tr('M_USER_RETENTION_PERIOD')}
              <Popup
                wide="very"
                position="bottom left"
                trigger={(
                  <Icon
                    name="info circle"
                    className={styles['popup-font']}
                  />
                )}
                content={(
                  <List bulleted>
                    {[0, 1, 2].map((l) => (
                      <List.Item className={styles['list-item']} key={l}>
                        <span>
                          {`${tr(`M_USER_RETENTION_${Object.keys(PeriodMap)[l].toUpperCase()}`)}：`}
                        </span>
                        {tr(`M_USER_RETENTION_MEMO_${l + 1}`)}
                      </List.Item>
                    ))}
                  </List>
                )}
              />
            </label>
            <Button.Group basic className="os-chart_quickSelect">
              {Object.keys(PeriodMap).map((p) => (
                <Button
                  active={searchForm.period === p}
                  content={PeriodMap[p].text}
                  data-period={p}
                  onClick={handleOnClickPeriod}
                  key={p}
                />
              ))}
            </Button.Group>
          </Form.Field>
          {/* 註冊日 */}
          <Form.Field
            label={tr('M_TEXT_CREATED_DAY')}
            control={DateRange}
            minDate={moment().tz('Etc/GMT+4').subtract(181, 'days')}
            startDate={(searchForm.start_at && moment(searchForm.start_at)) || ''}
            endDate={(searchForm.end_at && moment(searchForm.end_at)) || ''}
            setRange={handleCreatedDate}
            format={DF.date}
          />
          <Form.Button
            loading={loading}
            color="blue"
            content={tr('M_ACTION_SEARCH')}
            onClick={handleSubmit}
          />
        </Form.Group>
      </Form>

      {/* 登入留存率說明 */}
      <Menu secondary>
        <Menu.Menu className="right">
          <Button
            as="a"
            className={`item ${styles['menu-item']}`}
            onClick={handleExplanation}
          >
            <span>{tr('M_ANALYSIS_INDEX_DESCRIPTION')}</span>
            <Icon name={`${useActiveIcon} chevron`} />
          </Button>
        </Menu.Menu>
      </Menu>
      <div className={`${useActiveIcon} ${styles['explanation-block']}`}>
        <List>
          {[4, 5, 7].map((l) => (
            <List.Item as="li" key={l} className={styles['list-item']}>
              {tr(`M_USER_RETENTION_MEMO_${l}`)}
              {l === 4 && (
                <span className={styles['list-item-span']}>
                  {tr('M_USER_RETENTION_MEMO_8')}
                </span>
              )}
            </List.Item>
          ))}
        </List>
      </div>

      {/* Chart */}
      <RetentionRateLoginCharts
        // 圖表固定顯示"留存率(%)"，不會顯示留存人數
        data={data.total.percent}
        loading={loading}
        period={period}
      />

      {loading && <Loader active={loading} />}
      {!loading && (
        <TableOs
          tableParam={{ className: styles['customized-table'] }}
          columns={PeriodMap[period].columns}
          data={useTableData}
          leftElement={(
            <Button.Group>
              {Object.keys(ReportTypes).map((d) => (
                <Button
                  basic={d !== type}
                  content={ReportTypes[d].text}
                  color="blue"
                  data-type={d}
                  key={d}
                  onClick={handleIsPercentChange}
                />
              ))}
            </Button.Group>
          )}
        />
      )}

      <Pagination
        pagination={pagination}
        onPageChanged={getRetentionRateLogin}
      />
    </>
  );
};

export default RetentionRateLogin;
