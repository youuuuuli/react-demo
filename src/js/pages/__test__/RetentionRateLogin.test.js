import React from 'react';
import { render } from '@testing-library/react';
import RetentionRateLogin from '../RetentionRateLogin';

jest.mock('../../utils/DateRange');
jest.mock('../RetentionRateLoginCharts');

const setup = async () => {
  const utils = render(<RetentionRateLogin />);
  const { findByText } = utils;

  await findByText('2021-10-05 Tuesday');

  return utils;
};

describe('登入留存率', () => {
  it('基本', async () => {
    const { asFragment } = await setup();

    expect(asFragment()).toMatchSnapshot();
  });
});
