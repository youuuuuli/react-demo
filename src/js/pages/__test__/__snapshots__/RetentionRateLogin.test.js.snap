// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`登入留存率 基本 1`] = `
<DocumentFragment>
  <div
    class="page-title"
  />
  <form
    class="ui form search-wrap"
  >
    <div
      class="fields"
    >
      <div
        class="field"
      >
        <label>
          <i
            aria-hidden="true"
            class="info circle icon popup-font"
          />
        </label>
        <div
          class="ui basic buttons os-chart_quickSelect"
        >
          <button
            class="ui button"
            data-period="day"
          />
          <button
            class="ui active button"
            data-period="week"
          />
          <button
            class="ui button"
            data-period="month"
          />
        </div>
      </div>
      <div
        class="field"
      />
      <div
        class="field"
      >
        <button
          class="ui blue button"
        />
      </div>
    </div>
  </form>
  <div
    class="ui secondary menu"
  >
    <div
      class="menu right"
    >
      <a
        class="ui button item menu-item"
        role="button"
      >
        <span />
        <i
          aria-hidden="true"
          class="down chevron icon"
        />
      </a>
    </div>
  </div>
  <div
    class="down explanation-block"
  >
    <div
      class="ui list"
      role="list"
    >
      <li
        class="list-item"
        role="listitem"
      >
        <span
          class="list-item-span"
        />
      </li>
      <li
        class="list-item"
        role="listitem"
      />
      <li
        class="list-item"
        role="listitem"
      />
    </div>
  </div>
  <div
    class="ui basic clearing segment"
  >
    <div
      class="ui basic left floated segment"
    >
      <div
        class="ui buttons"
      >
        <button
          class="ui blue button"
          data-type="percent"
        />
        <button
          class="ui blue basic button"
          data-type="number"
        />
      </div>
    </div>
  </div>
  <table
    class="ui celled table customized-table"
  >
    <thead
      class=""
    >
      <tr
        class=""
      >
        <th
          class="center aligned"
        />
        <th
          class="center aligned"
        />
        <th
          class="center aligned"
        />
        <th
          class="center aligned"
        />
        <th
          class="center aligned"
        />
        <th
          class="center aligned"
        />
        <th
          class="center aligned"
        />
        <th
          class="center aligned"
        />
        <th
          class="center aligned"
        />
        <th
          class="center aligned"
        />
        <th
          class="center aligned"
        />
      </tr>
    </thead>
    <tbody
      class=""
    >
      <tr
        class=""
      >
        <td
          class="center aligned"
        >
          2021-10-05 Tuesday
        </td>
        <td
          class="center aligned"
        >
          <a
            class="ui button"
            data-created-date="2021-10-05"
            role="button"
          >
            1
          </a>
        </td>
        <td
          class="center aligned"
        >
          <span>
            0.00%
          </span>
        </td>
        <td
          class="center aligned"
        >
          <span>
            0.00%
          </span>
        </td>
        <td
          class="center aligned"
        >
          <span>
            0.00%
          </span>
        </td>
        <td
          class="center aligned"
        >
          <span>
            0.00%
          </span>
        </td>
        <td
          class="center aligned"
        />
        <td
          class="center aligned"
        />
        <td
          class="center aligned"
        />
        <td
          class="center aligned"
        />
        <td
          class="center aligned"
        />
      </tr>
      <tr
        class=""
      >
        <td
          class="center aligned"
        >
          2021-10-06 Wednesday
        </td>
        <td
          class="center aligned"
        >
          <a
            class="ui button"
            data-created-date="2021-10-06"
            role="button"
          >
            1
          </a>
        </td>
        <td
          class="center aligned"
        >
          <span>
            0.00%
          </span>
        </td>
        <td
          class="center aligned"
        >
          <span>
            0.00%
          </span>
        </td>
        <td
          class="center aligned"
        >
          <span>
            0.00%
          </span>
        </td>
        <td
          class="center aligned"
        >
          <span>
            0.00%
          </span>
        </td>
        <td
          class="center aligned"
        />
        <td
          class="center aligned"
        />
        <td
          class="center aligned"
        />
        <td
          class="center aligned"
        />
        <td
          class="center aligned"
        />
      </tr>
      <tr
        class=""
      >
        <td
          class="center aligned"
        >
          2021-10-07 Thursday
        </td>
        <td
          class="center aligned"
        >
          <a
            class="ui button"
            data-created-date="2021-10-07"
            role="button"
          >
            4
          </a>
        </td>
        <td
          class="center aligned"
        >
          <a
            class="ui button"
            data-created-date="2021-10-07"
            data-end-login-at="2021-10-14"
            data-start-login-at="2021-10-08"
            role="button"
          >
            25.00%
          </a>
        </td>
        <td
          class="center aligned"
        >
          <span>
            0.00%
          </span>
        </td>
        <td
          class="center aligned"
        >
          <a
            class="ui button"
            data-created-date="2021-10-07"
            data-end-login-at="2021-10-28"
            data-start-login-at="2021-10-22"
            role="button"
          >
            25.00%
          </a>
        </td>
        <td
          class="center aligned"
        >
          <span>
            0.00%
          </span>
        </td>
        <td
          class="center aligned"
        />
        <td
          class="center aligned"
        />
        <td
          class="center aligned"
        />
        <td
          class="center aligned"
        />
        <td
          class="center aligned"
        />
      </tr>
      <tr
        class=""
      >
        <td
          class="center aligned result-cell"
        />
        <td
          class="center aligned result-cell"
        >
          6
        </td>
        <td
          class="center aligned result-cell"
        >
          16.66%
        </td>
        <td
          class="center aligned result-cell"
        >
          0.00%
        </td>
        <td
          class="center aligned result-cell"
        >
          16.66%
        </td>
        <td
          class="center aligned result-cell"
        >
          0.00%
        </td>
        <td
          class="center aligned result-cell"
        >
          --
        </td>
        <td
          class="center aligned result-cell"
        >
          --
        </td>
        <td
          class="center aligned result-cell"
        >
          --
        </td>
        <td
          class="center aligned result-cell"
        >
          --
        </td>
        <td
          class="center aligned result-cell"
        >
          --
        </td>
      </tr>
      <tr
        class=""
      >
        <td
          class="center aligned result-cell"
        />
        <td
          class="center aligned result-cell"
        >
          6
        </td>
        <td
          class="center aligned result-cell"
        >
          16.66%
        </td>
        <td
          class="center aligned result-cell"
        >
          0.00%
        </td>
        <td
          class="center aligned result-cell"
        >
          16.66%
        </td>
        <td
          class="center aligned result-cell"
        >
          0.00%
        </td>
        <td
          class="center aligned result-cell"
        >
          --
        </td>
        <td
          class="center aligned result-cell"
        >
          --
        </td>
        <td
          class="center aligned result-cell"
        >
          --
        </td>
        <td
          class="center aligned result-cell"
        >
          --
        </td>
        <td
          class="center aligned result-cell"
        >
          --
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="ui fluid right aligned container"
  >
    <ul
      class="rc-pagination"
      data-testid="pagination"
      unselectable="unselectable"
    >
      <li
        class="rc-pagination-total-text"
      />
      <li
        aria-disabled="true"
        class="rc-pagination-disabled rc-pagination-prev"
        title="上一页"
      >
        <a
          class="rc-pagination-item-link"
          disabled=""
        />
      </li>
      <li
        class="rc-pagination-item rc-pagination-item-1 rc-pagination-item-active"
        tabindex="0"
        title="1"
      >
        <a>
          1
        </a>
      </li>
      <li
        aria-disabled="true"
        class="rc-pagination-disabled rc-pagination-next"
        title="下一页"
      >
        <a
          class="rc-pagination-item-link"
          disabled=""
        />
      </li>
      <li
        class="rc-pagination-options"
      >
        <div
          class="rc-pagination-options-size-changer rc-select rc-select-enabled"
        >
          <div
            aria-autocomplete="list"
            aria-controls="test-uuid"
            aria-expanded="false"
            aria-haspopup="true"
            class="rc-select-selection
            rc-select-selection--single"
            role="combobox"
            tabindex="0"
          >
            <div
              class="rc-select-selection__rendered"
            >
              <div
                class="rc-select-selection-selected-value"
                style="display: block; opacity: 1;"
                title="20 条/页"
              >
                20 条/页
              </div>
            </div>
            <span
              class="rc-select-arrow"
              style="user-select: none;"
              unselectable="on"
            >
              <i
                class="rc-select-arrow-icon"
              />
            </span>
          </div>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
